Problem 2.1 a

In [ ]:
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split


file_path = '/content/drive/My Drive/UniversalBank.csv'
data = pd.read_csv(file_path)

In [ ]:
data = pd.get_dummies(data, columns=['Education'])

In [ ]:
X = data.drop(columns=['ID', 'ZIP Code', 'Personal Loan'])
y = data['Personal Loan']

In [ ]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.4, random_state=42)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=1)

In [ ]:
new_customer = pd.DataFrame({
    'Age': [40],
    'Experience': [10],
    'Income': [84],
    'Family': [2],
    'CCAvg': [2],
    'Mortgage': [0],
    'Securities Account': [0],
    'CD Account': [0],
    'Online': [1],
    'CreditCard': [1],
    'Education_1': [0],  # Dummy variable for Education
    'Education_2': [1],  # Dummy variable for Education
    'Education_3': [0]   # Dummy variable for Education
})

In [ ]:
prediction = knn.predict(new_customer)

In [ ]:
if prediction[0] == 1:
    print("The customer is predicted to accept the loan offer.")
else:
    print("The customer is predicted to decline the loan offer.")

The customer is predicted to decline the loan offer.


B

In [ ]:
from sklearn.model_selection import cross_val_score

k_values = list(range(1, 21))

cv_scores = []
for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, X_train, y_train, cv=5, scoring='accuracy')
    cv_scores.append(scores.mean())

best_k = k_values[cv_scores.index(max(cv_scores))]
print("Best choice of k:", best_k)


Best choice of k: 5


C

In [ ]:

new_customer = pd.DataFrame({
    'Age': [40],
    'Experience': [10],
    'Income': [84],
    'Family': [2],
    'CCAvg': [2],
    'Mortgage': [0],
    'Securities Account': [0],
    'CD Account': [0],
    'Online': [1],
    'CreditCard': [1],
    'Education_1': [0],  # Dummy variable for Education
    'Education_2': [1],  # Dummy variable for Education
    'Education_3': [0]   # Dummy variable for Education
})

new_customer = new_customer[X.columns]


new_customer_scaled = scaler.transform(new_customer)

prediction_best = knn_best.predict(new_customer_scaled)
print("Prediction for the new customer (best k):", prediction_best[0])






Prediction for the new customer (best k): 0


2.2 A

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = data.drop(['ID', 'ZIP Code', 'Personal Loan'], axis=1)
y = data['Personal Loan']

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.4, random_state=42)


In [ ]:

train_data = pd.concat([X_train, y_train], axis=1)

acceptors = train_data[train_data['Personal Loan'] == 1]
non_acceptors = train_data[train_data['Personal Loan'] == 0]

# summary statistics for continuous predictors
continuous_predictors = ['Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Mortgage']
acceptors_continuous_stats = acceptors[continuous_predictors].describe().loc[['mean', 'std']]
non_acceptors_continuous_stats = non_acceptors[continuous_predictors].describe().loc[['mean', 'std']]

# percentages for categorical predictors
categorical_predictors = ['Securities Account', 'CD Account', 'Online', 'CreditCard', 'Education_1', 'Education_2', 'Education_3']
acceptors_categorical_stats = acceptors[categorical_predictors].mean() * 100
non_acceptors_categorical_stats = non_acceptors[categorical_predictors].mean() * 100

# Summary statistics
print("Summary statistics for continuous predictors (loan acceptors):\n", acceptors_continuous_stats)
print("\nSummary statistics for continuous predictors (non-acceptors):\n", non_acceptors_continuous_stats)
print("\nPercentages for categorical predictors (loan acceptors):\n", acceptors_categorical_stats)
print("\nPercentages for categorical predictors (non-acceptors):\n", non_acceptors_categorical_stats)


Summary statistics for continuous predictors (loan acceptors):
             Age  Experience      Income    Family     CCAvg    Mortgage
mean  45.261649   20.086022  144.025090  2.641577  3.962258  113.838710
std   11.732618   11.795612   31.999484  1.128693  2.108355  170.909835

Summary statistics for continuous predictors (non-acceptors):
             Age  Experience     Income    Family     CCAvg   Mortgage
mean  44.927600   19.685042  66.380742  2.353546  1.722080  52.259463
std   11.359641   11.350095  40.208764  1.149151  1.556032  93.428824

Percentages for categorical predictors (loan acceptors):
 Securities Account    12.544803
CD Account            28.673835
Online                62.724014
CreditCard            28.673835
Education_1           20.788530
Education_2           36.917563
Education_3           42.293907
dtype: float64

Percentages for categorical predictors (non-acceptors):
 Securities Account    10.363837
CD Account             3.785373
Online                59.6

2.2 B

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

data = pd.read_csv('UniversalBank.csv')

X = data.drop(['ID', 'ZIP Code', 'Personal Loan'], axis=1)
y = data['Personal Loan']

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.4, random_state=42)

categor_cols = ['Family', 'Education', 'Securities Account', 'CD Account', 'Online', 'CreditCard']
data = pd.get_dummies(data, columns=categor_cols, drop_first=True)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)

lda_model = LinearDiscriminantAnalysis()
lda_model.fit(X_train_scaled, y_train)

X_val_scaled = scaler.transform(X_val)
y_pred_val = lda_model.predict(X_val_scaled)

accuracy = accuracy_score(y_val, y_pred_val)
print(f'Accuracy on the validation set: {accuracy:.2%}')

# Confusion matrix
confus_matrix = confusion_matrix(y_val, y_pred_val)
print('\nConfusion Matrix:')
print(confus_matrix)

# Classification report
classif_report = classification_report(y_val, y_pred_val)
print('\nClassification Report:')
print(classif_report)

Accuracy on the validation set: 94.25%

Confusion Matrix:
[[1763   36]
 [  79  122]]

Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.98      0.97      1799
           1       0.77      0.61      0.68       201

    accuracy                           0.94      2000
   macro avg       0.86      0.79      0.82      2000
weighted avg       0.94      0.94      0.94      2000



Accuracy rate = 94.25%

False Positive (Type I Error): Times that the model predicted a personal loan, but the customer did not actually accept it. Here, there are 36 false positives.

False Negative (Type II Error): Times where the model predicted no personal loan, but the customer did accept it. Here, there are 79 false negatives.

The number of false negatives (79) is higher than that of false positives (36). Thus, the model will be more likely to make false negatives (Type II errors) than false positives (Type I errors) and one type of misclassification is more likely than the others.

